### PR TITLE
Attempt to improve vpc_subnet / vpc_net stability

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY

Over in community.aws we're seeing repeated failures due to API rate limits with the subnet/net changes (mostly waiter related)
Switch to a custom waiter that knows how to cope with the API Rate limits.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_vpc_subnet
ec2_vpc_net

##### ADDITIONAL INFORMATION

Will push changelog once tests complete

### CODE REVIEW

File: plugins/module_utils/waiters.py [hunk 1]
```diff
@@ -162,6 +162,19 @@
                 },
             ]
         },
+        "SubnetAvailable": {
```
Comment (alinabuzachis) [inline plugins/module_utils/waiters.py:165]: There seems a boto3 built-in waiter has already been implemented for doing this https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Waiter.VpcAvailable.
Does it makes sense to you?

File: plugins/module_utils/waiters.py [hunk 2]
```diff
@@ -250,6 +263,36 @@
                 },
             ]
         },
+        "VpcAvailable": {
```
Comment (alinabuzachis) [inline plugins/module_utils/waiters.py:266]: The same for doing this. https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Waiter.VpcAvailable
Does it makes sense to you?

File: plugins/module_utils/waiters.py [hunk 3]
```diff
@@ -514,6 +563,18 @@ def rds_model(name):
         core_waiter.NormalizedOperationMethod(
             ec2.describe_subnets
         )),
+    ('EC2', 'vpc_available'): lambda ec2: core_waiter.Waiter(
+        'vpc_available',
+        ec2_model('VpcAvailable'),
+        core_waiter.NormalizedOperationMethod(
+            ec2.describe_vpcs
+        )),
+    ('EC2', 'vpc_exists'): lambda ec2: core_waiter.Waiter(
```
Comment (alinabuzachis) [inline plugins/module_utils/waiters.py:572]: There is already a waiter for managing this as well https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Waiter.VpcExists
Does it make sense to you?

File: plugins/modules/ec2_vpc_net.py [hunk 4]
```diff
@@ -242,21 +243,32 @@ def vpc_exists(module, vpc, name, cidr_block, multi):
...
     except is_boto3_error_message('The functionality you requested is not available in this region.'):
         return False
 
 
-def get_vpc(module, connection, vpc_id):
+def wait_for_vpc_to_exist(module, connection, **params):
     # wait for vpc to be available
     try:
-        connection.get_waiter('vpc_available').wait(VpcIds=[vpc_id])
+        get_waiter(connection, 'vpc_exists').wait(**params)
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_net.py:257]: I would suggest adding this statement as well `except botocore.exceptions.WaiterError as e:`

File: plugins/modules/ec2_vpc_net.py [hunk 5]
```diff
@@ -242,21 +243,32 @@ def vpc_exists(module, vpc, name, cidr_block, multi):
...
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-        module.fail_json_aws(e, msg="Unable to wait for VPC {0} to be available.".format(vpc_id))
+        module.fail_json_aws(e, msg="Unable to wait for VPC creation.")
+
 
+def wait_for_vpc(module, connection, **params):
+    # wait for vpc to be available
+    try:
+        get_waiter(connection, 'vpc_available').wait(**params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_net.py:266]: I would suggest adding this statement as well `except botocore.exceptions.WaiterError as e:`

File: plugins/modules/ec2_vpc_net.py [hunk 6]
```diff
@@ -242,21 +243,32 @@ def vpc_exists(module, vpc, name, cidr_block, multi):
     return None
 
 
-@AWSRetry.backoff(delay=3, tries=8, catch_extra_error_codes=['InvalidVpcID.NotFound'])
 def get_classic_link_with_backoff(connection, vpc_id):
     try:
-        return connection.describe_vpc_classic_link(VpcIds=[vpc_id])['Vpcs'][0].get('ClassicLinkEnabled')
+        results = connection.describe_vpc_classic_link(aws_retry=True, VpcIds=[vpc_id])
+        return results['Vpcs'][0].get('ClassicLinkEnabled')
     except is_boto3_error_message('The functionality you requested is not available in this region.'):
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_net.py:250]: In addition to the boto3_error_message, what do you think about adding also `except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:`?

**Review by alinabuzachis — CHANGES_REQUESTED** [review]
Thank you @tremble for asking me to review.

File: plugins/module_utils/waiters.py (hunk 1 above)
Comment (tremble) [reply plugins/module_utils/waiters.py:165]: The trouble with the default boto3 waiter is that if it hits the API rate limiting errors it bombs straight out.  This custom waiter treats the rate limiting error as temporary and just retries after the next delay.

**Comment by tremble** [comment]
@alinabuzachis For what it's worth I did dig into what it would take to tune the botocore waiters

In theory we could do so by using `botocore.client.BaseClient._get_waiter_config()` (`client._get_waiter_config`) however, as an `_` method that leaves us reliant on a non-public interface, which I don't think would be a good idea.

This leaves is with copying the waiter configs we care about into waiters.py as the only sane way to inject retrying on API rate limits.

**Comment by alinabuzachis** [comment]
> @alinabuzachis For what it's worth I did dig into what it would take to tune the botocore waiters
> 
> In theory we could do so by using `botocore.client.BaseClient._get_waiter_config()` (`client._get_waiter_config`) however, as an `_` method that leaves us reliant on a non-public interface, which I don't think would be a good idea.
> 
> This leaves is with copying the waiter configs we care about into waiters.py as the only sane way to inject retrying on API rate limits.

@tremble Thank you for explaining. It makes sense to me. I am going to add a waiter even for NatGatewayAvailable since I used the built-in one. https://github.com/ansible-collections/community.aws/pull/445/files#diff-847118175f7fbd015daf612a3319157ca22874f706c01f26348f4f86f67384d9R706-R709

File: plugins/module_utils/waiters.py (hunk 1 above)
Comment (alinabuzachis) [reply plugins/module_utils/waiters.py:165]: @tremble Thank you for explaining. Hence, it makes sense using the custom one.

**Review by alinabuzachis — APPROVED** [review]

**Comment by tremble** [comment]
Change approved and Shippable tests passing.  Merging

**Pull request merged**